### PR TITLE
Refactor: Extract common utilities

### PR DESCRIPTION
**Pull request by DotSlash-A**
## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: MOE_EXPLANATION.md [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+
+Recombine with weights:
+Final = Σ(weight_i × expert_i_output) + shared_expert_output
+```
+
+---
+
+## 1. The MoEGate (Router) - How It Knows Which Expert to Call
+
+### Location: `deepseek_v2.py` lines 336-406
```
[the remark this answers is not in the document]
Comment (DotSlash-A) [reply MOE_EXPLANATION.md]: Deleted this file

File: MOE_EXPLANATION.md [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+For DeepSeek V2, it uses **grouped top-k** selection:
+
+1. **Grouped Top-K Logic** (`grouped_topk_gpu` function, lines 485-549):
+   ```python
+   # Step 1: Convert logits to probabilities
+   scores = torch.softmax(router_logits, dim=-1)  # (num_tokens, num_experts)
+   
+   # Step 2: Group experts and find best groups
+   # Experts are divided into groups (e.g., 64 experts in 8 groups = 8 experts per group)
+   group_scores = scores.view(num_token, num_expert_group, -1).max(dim=-1).values
```
[the remark this answers is not in the document]
Comment (DotSlash-A) [reply MOE_EXPLANATION.md]: deleted this file

File: python/sglang/srt/models/deepseek_common/utils.py [hunk 3]
```diff
@@ -21,3 +30,50 @@
...
+
+def add_forward_absorb_core_attention_backend(backend_name: str) -> None:
+    if backend_name not in FORWARD_ABSORB_CORE_ATTENTION_BACKENDS:
+        FORWARD_ABSORB_CORE_ATTENTION_BACKENDS.append(backend_name)
+        logger.info(f"Added {backend_name} to FORWARD_ABSORB_CORE_ATTENTION_BACKENDS.")
+
+
+def yarn_get_mscale(scale: float = 1, mscale: float = 1) -> float:
+    import math
+
```
[the remark this answers is not in the document]
Comment (DotSlash-A) [reply python/sglang/srt/models/deepseek_common/utils.py]: I made this change

File: test/manual/test_deepseek_model_loading.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
+"""
+Integration test: Load DeepSeek model and run inference.
+Requires: GPU with sufficient VRAM
+Run: python test/manual/test_deepseek_model_loading.py
+"""
+import torch
+from transformers import AutoTokenizer, AutoConfig
```
[the remark this answers is not in the document]
Comment (DotSlash-A) [reply test/manual/test_deepseek_model_loading.py]: I made this change

File: test/manual/test_deepseek_smoke.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        print(f"Attention backends count: {len(FORWARD_ABSORB_CORE_ATTENTION_BACKENDS)}")
+        print(f"CuBLAS >= 12.9: {_is_cublas_ge_129}")
+        
+        # Test enable_nextn_moe_bf16_cast_to_fp8
+        result = enable_nextn_moe_bf16_cast_to_fp8(None)
+        print(f"enable_nextn_moe_bf16_cast_to_fp8(None) = {result}")
+        
+    except Exception as e:
+        print(f"Function test failed: {e}")
+        import traceback
```
[the remark this answers is not in the document]
Comment (DotSlash-A) [reply test/manual/test_deepseek_smoke.py]: I made this change

File: test/manual/test_deepseek_utils_refactoring.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        
+        print("ALL TESTS PASSED!")
+        return 0
+        
+    except AssertionError as e:
+        print(f"\n TEST FAILED: {e}\n")
+        return 1
+    except Exception as e:
+        print(f"\n UNEXPECTED ERROR: {e}\n")
+        import traceback
```
[the remark this answers is not in the document]
Comment (DotSlash-A) [reply test/manual/test_deepseek_utils_refactoring.py]: I made this change
